### PR TITLE
Add support for local Wasm plugins

### PR DESCRIPTION
**Pull request by emcfarlane**
Add support for running custom plugins compiled with WebAssembly. The plugins must conform to the [WASI 0.1](https://wasi.dev/interfaces#wasi-01) specification. This system interface is supported by many languages. For [bufplugin-go](https://github.com/bufbuild/bufplugin-go/) based plugins see the [Go WASI docs](https://go.dev/wiki/WebAssembly#wasi-gooswasip1-port). Supporting plugins with the WASI specification allows the running of plugins using the pluginrpc protocol without any code modifications required by the user.

Declare a plugin with the suffix `.wasm` to use the Wasm runtime to invoke the plugin. For example the plugin [buf-plugin-field-option-safe-for-ml](check/internal/example/cmd/buf-plugin-field-option-safe-for-ml) can be compiled with `GOOS=wasip1 GOARCH=wasm` and added to your `buf.yaml` config:

```yaml
version: v2
plugins:
  - plugin: buf-plugin-field-option-safe-for-ml.wasm
```

### CODE REVIEW

File: private/usage/usage_unix.go [hunk 1]
```diff
@@ -12,8 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//go:build aix || darwin || dragonfly || freebsd || (js && wasm) || linux || netbsd || openbsd || solaris
-// +build aix darwin dragonfly freebsd js,wasm linux netbsd openbsd solaris
+//go:build unix || wasip1 || js
```
Comment (emcfarlane) [inline private/usage/usage_unix.go:15]: Use the unix build constraint and drop support for `+build` as that is required for go versions 1.16 and earlier.

File: private/bufpkg/bufpluginrunner/provider.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+			r.delegate,
+			// We know that Path is of at least length 1.
+			path[0],
+			path[1:]...,
+		), nil
+	case bufconfig.PluginConfigTypeLocalWasm:
+		if r.wasmRuntime == nil {
+			return nil, syserror.Newf("wasm runtime is required for local wasm plugins")
+		}
+		return newWasmRunner(
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/bufpkg/bufpluginrunner/provider.go:63]: They can't be implemented in pluginrpcutil as the pkg now depends on bufconfig for the `bufconfig.PluginConfig`. Could move pluginrpcutil under bufpkg?

[the remark this answers is not in the document]
Comment (bufdev) [reply private/bufpkg/bufpluginrunner/provider.go:63]: the `wasmRunner` in no way depends on `bufconfig.PluginConfig`, I'm not sure what you are talking about.

[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/bufpkg/bufpluginrunner/provider.go:63]: Referring to the `pluginrpcutil`. All good on moving the wasm runner.

File: private/buf/bufcli/cache.go [hunk 3]
```diff
@@ -135,6 +139,17 @@ func NewCommitProvider(container appext.Container) (bufmodule.CommitProvider, er
 	)
 }
 
+// CreatePluginCacheDir creates the cache directory for plugins.
+//
+// This is used by the [bufwasm.WithLocalCacheDir] option.
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/buf/bufcli/cache.go]: Removed `[]` use. It is supported in Godoc as a link but must be imported, so here it didn't link correctly.

File: private/buf/bufcli/cache.go [hunk 4]
```diff
@@ -103,6 +103,10 @@ var (
 	//
 	// Normalized.
 	v3CacheModuleLockRelDirPath = normalpath.Join("v3", "modulelocks")
+	// v3CachePluginsRelDirPath is the relative path to the plugins cache directory in its newest iteration.
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/buf/bufcli/cache.go]: Renamed and updated the docs to the `v3/wasmruntime` dir.

File: private/bufpkg/bufcheck/multi_client.go [hunk 5]
```diff
@@ -94,6 +95,7 @@ func (c *multiClient) Check(ctx context.Context, request check.Request) ([]*anno
 		jobs = append(
 			jobs,
 			func(ctx context.Context) error {
+				start := time.Now()
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/bufpkg/bufcheck/multi_client.go]: Switched to tracing.

File: private/bufpkg/bufpluginrunner/bufpluginrunner.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+package bufpluginrunner
+
+import (
+	"github.com/bufbuild/buf/private/bufpkg/bufconfig"
+	"github.com/bufbuild/buf/private/bufpkg/bufwasm"
+	"github.com/bufbuild/buf/private/pkg/command"
+	"pluginrpc.com/pluginrpc"
+)
+
+// RunnerProvider provides pluginrpc.Runners for a given plugin config.
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/bufpkg/bufpluginrunner/bufpluginrunner.go]: Explain on the interface the runner provider is by Type, with more detail on the `NewRunnerProvider` to what plugin types are supported.

[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/bufpkg/bufpluginrunner/bufpluginrunner.go]: Fixed.

File: private/bufpkg/bufpluginrunner/provider.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/bufpkg/bufpluginrunner/provider.go]: File moved back to `bufcheck/runner_provider.go`.

File: private/bufpkg/bufpluginrunner/provider.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+func newRunnerProviderOptions() *runnerProviderOptions {
+	return &runnerProviderOptions{
+		wasmRuntime: bufwasm.NewUnimplementedRuntime(),
+	}
+}
+
+// wasmRunner is a runner that loads a Wasm plugin.
+type wasmRunner struct {
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/bufpkg/bufpluginrunner/provider.go]: Moved to [pluginrpcutil/wasm_runner.go](https://github.com/bufbuild/buf/pull/3349/files#diff-d45e4483ed43229206010e22dbde1b87f40c639bb83b0209e94994f9dac8d58d)

File: private/bufpkg/bufwasm/bufwasm.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/bufpkg/bufwasm/bufwasm.go]: Moved to `pkg/wasm`

File: private/bufpkg/bufwasm/bufwasm.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+// Package bufwasm provides a Wasm runtime for plugins.
+package bufwasm
+
+import (
+	"context"
+
+	"pluginrpc.com/pluginrpc"
+)
+
+// Plugin is a Wasm module.
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/bufpkg/bufwasm/bufwasm.go]: Originally thought that Wasm Plugin as its a Wasm module that implements the pluginrpc spec. Have renamed to CompiledModule.

File: private/buf/bufcli/cache.go [hunk 11]
```diff
@@ -135,6 +139,17 @@ func NewCommitProvider(container appext.Container) (bufmodule.CommitProvider, er
 	)
 }
 
+// CreatePluginCacheDir creates the cache directory for plugins.
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/buf/bufcli/cache.go]: Renamed to `CreateWasmRuntimeCacheDir`.

File: private/bufpkg/bufwasm/bufwasm.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (bufdev) [reply private/bufpkg/bufwasm/bufwasm.go]: Module also has specific meaning in Buf.

File: CHANGELOG.md [hunk 12]
```diff
@@ -2,7 +2,7 @@
 
 ## [Unreleased]
 
-- No changes yet.
+- Add support for custom WebAssembly plugins for lint and breaking changes plugins. Use the `.wasm` file extension to specify a WebAssembly plugin.
```
Comment (bufdev) [inline CHANGELOG.md:5]: On line 13, we refer to this as "custom lint and breaking change plugins", however here you refer to them as "lint and breaking changes plugins". The rest of the file also wraps at 100 columns.

File: private/bufpkg/bufcheck/bufcheck.go [hunk 13]
```diff
@@ -165,26 +165,25 @@ type RunnerProvider interface {
...
-				path[1:]...,
-			), nil
-		},
-	)
+// NewRunnerProvider returns a new RunnerProvider for the command.Runner and wasm.Runtime.
+//
+// This implementation should only be used for local applications. It is safe to
+// use concurrently.
+//
+// The RunnerProvider selects the correct runner based on the PluginConfig.Type.
```
Comment (bufdev) [inline private/bufpkg/bufcheck/bufcheck.go:179]: `Runner`

File: private/bufpkg/bufwasm/bufwasm.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (bufdev) [reply private/bufpkg/bufwasm/bufwasm.go]: Nevermind on this - `CompiledModule` seems to be a thing for Wasm too

File: private/pkg/wasm/wasm.go [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+	Release(ctx context.Context) error
+}
+
+// NewRuntime creates a new Wasm runtime.
+func NewRuntime(ctx context.Context, options ...RuntimeOption) (Runtime, error) {
+	return newRuntime(ctx, options...)
+}
+
+// RuntimeOption is an option for NewRuntime.
+type RuntimeOption interface {
```
Comment (bufdev) [inline private/pkg/wasm/wasm.go:55]: `type RuntimeOption func(*runtimeOptions)`, please follow the same pattern as other packages.

File: private/pkg/wasm/wasm.go [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+// Package wasm provides a Wasm runtime for plugins.
+package wasm
+
+import (
+	"context"
+
+	"pluginrpc.com/pluginrpc"
+)
+
+// CompiledModule is a Wasm module ready to be run.
```
Comment (bufdev) [inline private/pkg/wasm/wasm.go:24]: It feels like this should be called `Runner`

File: private/pkg/wasm/runtime.go [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+		cache, err = wazero.NewCompilationCacheWithDir(cfg.cacheDir)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create compilation cache: %w", err)
+		}
+		runtimeConfig = runtimeConfig.WithCompilationCache(cache)
+	}
+	r := wazero.NewRuntimeWithConfig(ctx, runtimeConfig)
+
+	// Init wasi.
+	wasi_snapshot_preview1.MustInstantiate(ctx, r)
```
Comment (bufdev) [inline private/pkg/wasm/runtime.go:64]: This seems to be instantiating a global, and doing so with a potential panic, can you elaborate?

File: private/pkg/wasm/runtime.go [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+	var cache wazero.CompilationCache
+	if cfg.cacheDir != "" {
+		var err error
+		cache, err = wazero.NewCompilationCacheWithDir(cfg.cacheDir)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create compilation cache: %w", err)
+		}
+		runtimeConfig = runtimeConfig.WithCompilationCache(cache)
+	}
+	r := wazero.NewRuntimeWithConfig(ctx, runtimeConfig)
```
Comment (bufdev) [inline private/pkg/wasm/runtime.go:61]: We don't use single character variable names in this codebase beyond method receivers

File: private/pkg/wasm/wasm.go (hunk 15 above)
Comment (emcfarlane) [reply private/pkg/wasm/wasm.go:24]: From this comment https://github.com/bufbuild/buf/pull/3349#discussion_r1784915200 did you mean its fine to leave as a `CompiledModule`? The `Runner` for Wasm modules is provided by `pluginrpcutil.NewWasmRunner`. The Wasm module is runnable but it's not a generic runner of plugins.

Comment (bufdev) [reply private/pkg/wasm/wasm.go:24]: I did mean that initially, but this *is* factually a runner, so this should probably be called `Runner` itself as well. Another option would be `RunCloser`, which would match ie `ReadCloser, WriteCloser` in being a union of `Runner` and `Closer`, but it's a little weird as this `Close` requires a `context.Context`.

I think with all that said though - I guess yes, you can keep it as `CompiledModule` in this case.

File: private/pkg/wasm/runtime.go (hunk 16 above)
Comment (emcfarlane) [reply private/pkg/wasm/runtime.go]: Fixed and improved comment.

File: private/pkg/wasm/wasm.go [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+//
+// It is safe to use this module concurrently. When done, call Release to free
+// resources held by the CompiledModule. All CompiledModules created by the
+// same Runtime will be invalidated when the Runtime is released.
+//
+// Memory is limited by the runtime. To restrict CPU usage, cancel the context.
+type CompiledModule interface {
+	pluginrpc.Runner
+	// Release releases all resources held by the compiled module.
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply private/pkg/wasm/wasm.go]: Renamed to Close.

File: private/pkg/wasm/wasm.go [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+//
+// This option is only safe use in CLI environments.
+func WithLocalCacheDir(cacheDir string) RuntimeOption {
+	return func(runtimeOptions *runtimeOptions) {
+		runtimeOptions.cacheDir = cacheDir
+	}
+}
+
+// UnimplementedRuntime is an unimplemented Runtime.
+var UnimplementedRuntime = unimplementedRuntime{}
```
Comment (bufdev) [inline private/pkg/wasm/wasm.go:81]: global vars always go at the top, just below global consts

**Comment by bufdev** [comment]
I just merged a PR that will conflict with CHANGELOG.md, you'll need to merge

**Review by bufdev — APPROVED** [review]

**Pull request merged**